api/src/main/java/org/openmrs/module/datafilter/impl/ImplDataFilterListener.java

```diff
@@ -86,19 +87,21 @@ public boolean onEnableFilter(DataFilterContext filterContext) {
 			filterContext.setParameter(ImplConstants.PARAM_NAME_ROLES, roles);
 			
 		} else if (filterContext.getFilterName().startsWith(ImplConstants.PROGRAM_BASED_FILTER_NAME_PREFIX)) {
-			Collection<String> userProgramRoleNames = new HashSet();
+			// Avoid a 'select IN ()' which would be an invalid query, in theory we expect no role to match #####
+			Collection<String> userProgramRoleNames = new HashSet<>(Arrays.asList("#####"));
+			
 			Collection<String> allProgramRoleNames = AccessUtil.getAllProgramRoles();
+			if (allProgramRoleNames.isEmpty()) {
+				//Avoid a 'select IN ()' which would be an invalid query, in theory we expect no role to match #####
+				allProgramRoleNames.add("#####");
+			}
+			
 			if (Context.isAuthenticated()) {
 				Collection<Role> userProgramRoles = Context.getAuthenticatedUser().getAllRoles().stream()
 				        .filter(r -> allProgramRoleNames.contains(r.getName())).collect(Collectors.toList());
 				
 				userProgramRoleNames = userProgramRoles.stream().map(r -> r.getName()).collect(Collectors.toSet());
 				
-				if (allProgramRoleNames.isEmpty()) {
```

Comment:
I think this if clause and the one after it just need to be both moved out of their enclosing if clause to happen after and not before like you are doing

Reply:
And I think you need a context sensitive test to verify the fix, this should be a simple test to put together.

Reply:
Yes but not when depending on the underlying H2 database env.

Reply:
It is just a test like any other context sensitive test, I don't see what the H2 database has do to with it, please explain

Reply:
@wluyima (I'm just being the messenger here), @Ruhanga is reporting that the faulty SQL syntax such as:
```sql
SELECT IN ()
```
* fails with MySQL/MariaDB;
* is in fact ok with H2.

Nathan is touching on that [here](https://github.com/openmrs/openmrs-module-datafilter/pull/19/files#r1045446713) already.

What do you recommend doing?

Reply:
Got it! Yes H2 provides an option to run it in a mode of a specific database type in this case MySQL and you can do it for a single test class

Reply:
The H2 connection URL should be tweaked in the end by providing the url to the database in question. See [here](http://www.h2database.com/html/features.html),
>To use the MySQL mode, use the database URL jdbc:h2:~/test;MODE=MySQL;DATABASE_TO_LOWER=TRUE. When case-insensitive identifiers are needed append ;CASE_INSENSITIVE_IDENTIFIERS=TRUE to URL. Do not change value of DATABASE_TO_LOWER after creation of database.

This will still affect other classes in the end.

Reply:
No it won't, have you tried it? Unless there are other existing invalid queries that don't work on MySQL but work on H2 which would be a good thing to discover.

In any case even if it did, it would be the better option because running tests against H2 is technically wrong since no one uses H2 in production, with the invention of tools like docker and embedded MySQL which didn't exist at the time when the test framework was being put together, otherwise using H2 masks invalid queries errors in tests which would otherwise arise in prod.

I have always wanted to advocate for running context sensitive tests in MySQL mode by default, preferably we need a ticket to modify core to stop using H2 in tests.

Reply:
FWIW OpenMRS platform 2.6.0 supports running these tests on mysql or postgresql. https://issues.openmrs.org/browse/TRUNK-6028

Reply:
Glad to hear that @dkayiwa ! +10000

Reply:
For the dbsync and OpenMRS EIP projects I specifically used test containers for this reason

Reply:
Yes for sure, I've done this with mekomsolutions/commonreports as well and shouldn't be a big deal to get it through, only that this will require not minor test changes as the fix is.

Reply:
So looks like everyone is going in for the fix with as close to runtime env as possible. So I'll introduce the MySQL testcontainer for tests to run on.

Reply:
You don't need to set DATABASE_TO_LOWER

Reply:
> I think this if clause and the one after it just need to be both moved out of their enclosing if clause to happen after and not before like you are doing

@Ruhanga did you also see this comment?

Reply:
>did you also see this comment?

Yes @wluyima.

>You don't have to set DATABASE_TO_LOWER

Ohh, ok sure. I was just demonstrating an example snippet. FWIW, simply setting the MODE without providing the corresponding databse url yields no difference.

Reply:
That's incorrect

Reply:
I'm trying to imply that appending `MODE=MySQL` to the H2 overall URL requires an existing MySQL db sub url to work as expected. I.e. instead of 
```
jdbc:h2:mem:openmrs;DB_CLOSE_DELAY=30;LOCK_TIMEOUT=10000;MODE=MySQL
```
I have to do some thing along the following line
```
jdbc:h2:localhost:<DATABASE_PORT>/openmrs?autoReconnect=true&sessionVariables=default_storage_engine%3DInnoDB&useUnicode=true&characterEncoding=UTF-8;MODE=MySQL
```

Reply:
I'm trying to imply that appending `MODE=MySQL` to the H2 overall URL requires an existing MySQL db sub url to work as expected. I.e. instead of 
```
jdbc:h2:mem:openmrs;DB_CLOSE_DELAY=30;LOCK_TIMEOUT=10000;MODE=MySQL
```
I have to do some thing along the following line
```
jdbc:h2:localhost:<DATABASE_PORT>/openmrs?autoReconnect=true&sessionVariables=default_storage_engine%3DInnoDB&useUnicode=true&characterEncoding=UTF-8;MODE=MySQL
```

Reply:
I've produced a test demonstrating thrown error here https://github.com/openmrs/openmrs-module-datafilter/pull/20. This uncovers bugs and breaks ~50% of the existing tests whose errors where silent with an H2 db env.